admin/controllers/school.js

```diff
@@ -486,19 +511,52 @@ const savePupilNotTakingCheck = async (req, res, next) => {
     if (!savedPupils) {
       return next(new Error('Cannot save pupils'))
     }
+    flashMessage = `${savedPupils.length} pupil reasons updated`
+    req.flash('info', flashMessage)
   } catch (error) {
     return next(error)
   }
 
-  // @TODO: list of pupils with reasons
-  let pupilsList = []
+  // Get attendance code index
+  try {
+    attendanceCodes = await AttendanceCode.getAttendanceCodes().exec()
+  } catch (error) {
+    console.log('ERROR getting attendance codes', error)
+    return next(error)
+  }
+
+  let pupils = await fetchPupilsWithReasons(req.user.School)
+  pupilsList = await Promise.all(pupils.map(async (p) => {
+    if (p.attendanceCode !== undefined && p.attendanceCode._id !== undefined) {
+      let accCode = attendanceCodes.filter(ac => JSON.stringify(ac._id) === JSON.stringify(p.attendanceCode._id))
+      p.reason = accCode[0].reason
+      p.highlight = (postedPupils.filter(pp => JSON.stringify(pp) === JSON.stringify(p._id))).length > 0
+    }
+    return p
+  }))
 
   return res.render('school/save-pupils-not-taking-check', {
     breadcrumbs: req.breadcrumbs(),
-    pupilsList
+    pupilsList,
+    messages: req.flash('info')
   })
 }
 
+const removePupilNotTakingCheck = async (req, res, next) => {
+  if (req.params.pupilId === undefined) {
```

Comment:
Unless you expect the pupilId to ever be 0,empty array or object or nul this can just be simplified as:
`if (req.params.pupild) {`